notebooks/machine_learning/raw/ex8.ipynb

```diff
@@ -0,0 +1,244 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Get predictions from the trained model using the validation features\n",
+    "pred_y = ____\n",
+    "\n",
+    "# Calculate the accuracy of the trained model with the validation targets and predicted targets\n",
+    "accuracy = ____\n",
+    "\n",
+    "print(f\"Accuracy: {accuracy}\")\n",
+    "\n",
+    "step_2.check()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The lines below will show you a hint or the solution.\n",
+    "#step_2.hint()\n",
+    "#step_2.solution()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#%%RM_IF(PROD)%%\n",
+    "# Get predictions from the trained model using the validation features\n",
+    "pred_y = model.predict(val_X)\n",
+    "\n",
+    "# Calculate the accuracy of the trained model with the validation targets and predicted targets\n",
+    "accuracy = metrics.accuracy_score(val_y, pred_y)\n",
+    "\n",
+    "print(\"Accuracy, Random Forest Classifier: {:.3f}\".format(accuracy))\n",
+    "step_2.assert_check_passed()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Interpret the results\n",
+    "\n",
+    "Finally, calculate the confusion matrix for the classifier. We'll also normalize the confusion matrix to get the true positive rates."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "confusion = metrics.confusion_matrix(val_y, pred_y)\n",
+    "print(f\"Confusion matrix:\\n{confusion}\")\n",
+    "\n",
+    "# Normalizing by the true label counts to get rates\n",
+    "print(f\"\\nNormalized confusion matrix:\")\n",
+    "for row in confusion:\n",
+    "    print(row / row.sum())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Looking at the true positive rate, do you think the model is doing well at classifying pulsars from radio wave signals? Is the model misclassifying noise as pulsars or missing pulsars in the data?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#step_3.solution()"
+   ]
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "Python 3",
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.6.8"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```

Comment:
This exercise is pretty good. 

It'd be nice to add another ThoughtQuestion which is more abstract or conceptual.
We've heard a good response on those types of questions.  One example would be

"Think of a circumstance where you have an accuracy of 99%, and that seems surprisingly low"

You could then cut the imbalanced stuff from what the user passively reads... and give an answer to this thought question
```It would be surprising to have only 99% accuracy if a single outcome happens more than 99% of the time.
For example, if you are predicting whether a credit card transaction is fraudulent, and 99.9% of credit card transactions are legitimate, then you could get 99.9% accuracy by predicting everything is legitimate, even without a model.```

A user will remember this better if it's the response to a puzzle they've thought about, rather than it being stuck in the middle of a tutorial they are passively reading through."